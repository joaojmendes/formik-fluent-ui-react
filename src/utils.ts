import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
