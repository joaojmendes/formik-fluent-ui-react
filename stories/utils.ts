import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
