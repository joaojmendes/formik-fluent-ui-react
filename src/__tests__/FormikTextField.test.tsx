import { Field, FieldProps, Form, Formik } from 'formik'
import { TextField } from 'office-ui-fabric-react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import {
  FormikMaskedTextField,
  FormikTextField,
  mapFieldToTextField,
} from '../FormikTextField'
import { noop, serialize } from './utils'

class Values {
  public email: string = 'foo@bar.com'
}

function createFieldProps(
  value: string = 'foo@bar.com',
  errorMessage?: string | undefined
): FieldProps<Values> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'email',
    },
    form: {
      setFieldValue: jest.fn(),
      handleBlur: jest.fn(),
      errors: {
        email: errorMessage,
      },
      touched: {
        email: !!errorMessage,
      },
    },
  } as any
}

test('<FormikTextField /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field name="email" label="Email" component={FormikTextField} />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikTextField /> renders a Fabric <TextField />', () => {
  const fieldProps = createFieldProps()

  const formikTextField = renderer.create(
    <FormikTextField {...fieldProps} label="Email" />
  )
  const fabricTextField = renderer.create(
    <TextField {...mapFieldToTextField(fieldProps)} label="Email" />
  )
  expect(serialize(formikTextField)).toBe(serialize(fabricTextField))
})

test('mapFieldToTextField() maps FieldProps to ITextFieldProps', () => {
  const { field, form } = createFieldProps()
  const props = mapFieldToTextField({ form, field })

  expect(props.value).toBe(field.value)

  props.onBlur!(null as any)

  expect(field.onBlur).toHaveBeenCalledTimes(1)

  props.onChange!(null as any, 'bar')

  expect(field.onChange).toHaveBeenCalledTimes(1)
  expect(field.onChange).toHaveBeenCalledWith(null, 'bar')
})

test('mapFieldToTextField() additionally maps error message only when touched', () => {
  const errorMessage = 'not a valid email'
  const { field, form } = createFieldProps(new Values().email, errorMessage)

  const props = mapFieldToTextField({
    form: {
      ...form,
      touched: { [field.name]: false },
    },
    field,
  })

  expect(props.errorMessage).not.toBe(errorMessage)

  const propsTouched = mapFieldToTextField({
    form,
    field,
  })

  expect(propsTouched.errorMessage).toBe(errorMessage)
})

test('allow errorMessage prop overwrite', () => {
  const errorMessage = 'not a valid email'
  const { field, form } = createFieldProps(new Values().email, errorMessage)

  const formikTextField = renderer.create(
    <FormikTextField {...{ field, form }} errorMessage={'overwrite'} />
  )
  const instance = formikTextField.root

  expect(instance).not.toBeNull()
  expect(() =>
    instance!.findByProps({ errorMessage: 'overwrite' })
  ).not.toThrow()
})

test('<FormikMaskedTextField /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={{ masked: '' }} onSubmit={noop}>
      <Form>
        <Field
          name="masked"
          label="Masked"
          mask="(999) 999 - 9999"
          component={FormikMaskedTextField}
        />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})
