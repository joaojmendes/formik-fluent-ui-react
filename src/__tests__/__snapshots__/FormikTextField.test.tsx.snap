// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikMaskedTextField /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ms-TextField root-48"
  >
    <div
      className="ms-TextField-wrapper wrapper-49"
    >
      <label
        className="ms-Label root-57"
        htmlFor="TextField9"
      >
        Masked
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup-50"
      >
        <input
          aria-invalid={false}
          className="ms-TextField-field field-51"
          id="TextField9"
          name="masked"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          onPaste={[Function]}
          type="text"
          value="(___) ___ - ____"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`<FormikTextField /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ms-TextField root-48"
  >
    <div
      className="ms-TextField-wrapper wrapper-49"
    >
      <label
        className="ms-Label root-57"
        htmlFor="TextField1"
      >
        Email
      </label>
      <div
        className="ms-TextField-fieldGroup fieldGroup-50"
      >
        <input
          aria-invalid={false}
          className="ms-TextField-field field-51"
          id="TextField1"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          type="text"
          value="foo@bar.com"
        />
      </div>
    </div>
  </div>
</form>
`;
